src/dune_trace/out.ml

```diff
@@ -1,16 +1,112 @@
 open Stdune
 
 type t =
-  { out : out_channel
+  { fd : Unix.file_descr
+  ; mutable buf : Bigstringaf.t
+  ; mutable pos : int
   ; cats : Category.Set.t
   }
 
-(* all fields of record used *)
+(* CR-someday rgrinberg: remove this once we drop support for < 5.2 *)
+external write_bigstring
+  :  Unix.file_descr
+  -> Bigstringaf.t
+  -> int
+  -> int
+  -> int
+  = "dune_trace_write"
 
-let close t = close_out t.out
-let create cats out = { out; cats = Category.Set.of_list cats }
-let flush t = flush t.out
-let emit t event = Csexp.to_channel t.out event
+let flush =
+  (* This loop will almost always result in a single write, but we make sure to
+     write everything (albeit inefficiently) if the user is running out of disk
+     space, is on NFS, or some exotic operation system that doesn't give us
+     atomic writes with [O_APPEND] *)
+  let rec loop t pos len =
+    if len = 0
+    then t.pos <- 0
+    else (
+      match write_bigstring t.fd t.buf pos (t.pos - pos) with
+      | n -> loop t (pos + n) (len - n)
+      | exception e ->
+        let () =
+          (* inefficient, but we just want to make sure we're preserving the
+             invariants of [t] even if we fail for some odd reason. *)
+          let buf = Bigstringaf.create (Bigstringaf.length t.buf) in
+          Bigstringaf.blit buf ~src_off:pos t.buf ~dst_off:0 ~len;
+          t.buf <- buf
+        in
+        raise e)
+  in
+  fun t -> loop t 0 t.pos
+;;
+
+let close t =
+  flush t;
+  Unix.close t.fd
+;;
+
+let create cats path =
+  let fd =
+    Unix.openfile
+      (Path.to_string path)
+      [ O_TRUNC; O_APPEND; O_CLOEXEC; O_WRONLY; O_CREAT ]
+      0o777
+  in
+  let cats = Category.Set.of_list cats in
+  let buf = Bigstringaf.create (1 lsl 16) in
+  { fd; cats; pos = 0; buf }
+;;
+
+let to_buffer t sexp =
+  let rec loop = function
+    | Csexp.Atom str ->
+      let len = String.length str in
+      let () =
+        let len_str = string_of_int len in
+        Bigstringaf.blit_from_string
+          len_str
+          t.buf
+          ~dst_off:t.pos
+          ~src_off:0
+          ~len:(String.length len_str);
+        t.pos <- t.pos + String.length len_str
+      in
+      Bigstringaf.set t.buf t.pos ':';
+      t.pos <- t.pos + 1;
+      Bigstringaf.blit_from_string
+        str
+        t.buf
+        ~dst_off:t.pos
+        ~src_off:0
+        ~len:(String.length str);
+      t.pos <- t.pos + len
+    | List e ->
+      Bigstringaf.set t.buf t.pos '(';
+      t.pos <- t.pos + 1;
+      List.iter ~f:loop e;
+      Bigstringaf.set t.buf t.pos ')';
+      t.pos <- t.pos + 1
+  in
+  loop sexp
+;;
+
+let emit_buffered t event =
+  let needed = Csexp.serialised_length event in
+  let buf_len = Bigstringaf.length t.buf in
+  let available = buf_len - needed in
+  if available < needed
```

Comment:
This doesn't make sense to me. Shouldn't we just get rid of `available` and compare `needed` and `buf_len` directly instead?